crates/sui/src/upgrade_compatibility.rs

```diff
@@ -1301,6 +1304,58 @@ fn struct_ability_mismatch_diag(
     Ok(diags)
 }
 
+fn field_to_string(field: &Field) -> (String, String, String) {
+    let mut field_full = format!("'{}: {}'", field.name, field.type_);
+    let mut field_name = format!("'{}'", field.name);
+    let field_type = format!("'{}'", field.type_);
+
+    if let Some(pos_num) = Regex::new(r"^pos(\d)+$")
+        .ok()
+        .and_then(|r| r.captures(field.name.as_str()))
+        .map(|c| c.get(1).unwrap().as_str())
+    {
+        field_name = format!("at position {}", pos_num);
+        field_full = format!("{} {}", field_type, field_name);
+    }
+
+    (field_full, field_name, field_type)
+}
+
+use regex::Regex;
+/// returns a message for the given field
+fn field_mismatch_message(old_field: &Field, new_field: &Field) -> (Declarations, String) {
+    let (old_field_full, old_field_name, old_field_type) = field_to_string(old_field);
+    let (new_field_full, new_field_name, new_field_type) = field_to_string(new_field);
+
+    match (
```

Comment:
why does this match not include the pair `old_field_full` / `new_field_full`?

Reply:
seems better to rely on the &Field type equality to do this otherwise we might accidentally lose information in the field_to_string if the underlying implementation changes for either how equality is checked or how `ToString` works.